crates/aptos-crypto/src/webauthn/webauthn_p256_keys.rs

```diff
@@ -0,0 +1,280 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This file implements traits for WebAuthn P256 private keys and public keys.
+
+#[cfg(any(test, feature = "fuzzing"))]
+use crate::test_utils::{self, KeyPair};
+use crate::{
+    hash::CryptoHash,
+    p256::{P256PrivateKey, P256PublicKey, P256Signature, P256_PUBLIC_KEY_LENGTH},
+    traits::*,
+    webauthn::webauthn_p256_sigs::WebAuthnP256Signature,
+};
+use aptos_crypto_derive::{DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
+use core::convert::TryFrom;
+use p256::{self, ecdsa};
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest::prelude::*;
+use serde::Serialize;
+use std::fmt;
+
+/// A WebAuthn P256 private key
+#[derive(DeserializeKey, SerializeKey, SilentDebug, SilentDisplay)]
+pub struct WebAuthnP256PrivateKey(pub(crate) P256PrivateKey);
+
+impl private::Sealed for WebAuthnP256PrivateKey {}
+
+#[cfg(feature = "assert-private-keys-not-cloneable")]
+static_assertions::assert_not_impl_any!(WebAuthnP256PrivateKey: Clone);
+
+#[cfg(any(test, feature = "cloneable-private-keys"))]
+impl Clone for WebAuthnP256PrivateKey {
+    fn clone(&self) -> Self {
+        let serialized: &[u8] = &(self.to_bytes());
+        WebAuthnP256PrivateKey::try_from(serialized).unwrap()
+    }
+}
+
+/// A WebAuthn P256 public key
+#[derive(DeserializeKey, Clone, SerializeKey)]
+pub struct WebAuthnP256PublicKey(pub(crate) P256PublicKey);
+
+impl private::Sealed for WebAuthnP256PublicKey {}
+
+impl WebAuthnP256PrivateKey {
+    /// The length of the WebAuthnP256PrivateKey
+    pub const LENGTH: usize = P256PrivateKey::LENGTH;
+
+    /// Serialize a WebAuthnP256PrivateKey.
+    pub fn to_bytes(&self) -> [u8; P256PrivateKey::LENGTH] {
+        self.0.to_bytes()
+    }
+
+    /// Deserialize an WebAuthnP256PrivateKey without any validation checks apart from expected key size.
+    fn from_bytes_unchecked(
+        bytes: &[u8],
+    ) -> std::result::Result<WebAuthnP256PrivateKey, CryptoMaterialError> {
+        match p256::ecdsa::SigningKey::from_slice(bytes) {
+            Ok(p256_secret_key) => Ok(WebAuthnP256PrivateKey(P256PrivateKey(p256_secret_key))),
+            Err(_) => Err(CryptoMaterialError::DeserializationError),
+        }
+    }
+
+    /// This may be problematic
+    /// Trait method is required but DOES NOT work well for WebAuthn signatures
+    /// because verification_data is unknown without auth_data and client_data_json
+    ///
+    /// Private function aimed at minimizing code duplication between sign
+    /// methods of the SigningKey implementation. This should remain private.
+    fn sign_arbitrary_message(&self, message: &[u8]) -> WebAuthnP256Signature {
```

Comment:
Why can't this just return a P256Signature instead?

Reply:
Yeah it might be possible, but the trait implementation on `WebAuthnP256PrivateKey` for `SigningKey` relies on this `sign_arbitrary_message` returning a `WebAuthnP256Signature`. Here's the code below for reference

```rust
impl SigningKey for WebAuthnP256PrivateKey {
    type VerifyingKeyMaterial = WebAuthnP256PublicKey;
    type SignatureMaterial = WebAuthnP256Signature;

    fn sign<T: CryptoHash + Serialize>(
        &self,
        message: &T,
    ) -> Result<WebAuthnP256Signature, CryptoMaterialError> {
        Ok(WebAuthnP256PrivateKey::sign_arbitrary_message(
            self,
            signing_message(message)?.as_ref(),
        ))
    }

    #[cfg(any(test, feature = "fuzzing"))]
    fn sign_arbitrary_message(&self, message: &[u8]) -> WebAuthnP256Signature {
        WebAuthnP256PrivateKey::sign_arbitrary_message(self, message)
    }
}
```